### PR TITLE
Adding new burner contract

### PR DESCRIPTION
**Pull request by jamesmorgan**
Subgraph - https://github.com/knownorigin/known-origin-subgraph/pull/11

### CODE REVIEW

File: contracts/v2/artist-controls/ArtistEditionBurner.sol [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  /**
+   * @dev Sets the provided edition to either a deactivated state or reduces the available supply to zero
+   * @dev Only callable from edition artists defined in KODA NFT contract
+   * @dev Only callable when contract is not paused
+   * @dev Reverts if edition is invalid
+   * @dev Reverts if edition is not active in KDOA NFT contract
+   */
+  function deactivateOrReduceEditionSupply(uint256 _editionNumber) external whenNotPaused {
+    (address artistAccount, uint256 _) = kodaAddress.artistCommission(_editionNumber);
+    require(msg.sender == artistAccount || msg.sender == owner, "Only from the edition artist account");
```
Comment (andygray) [inline contracts/v2/artist-controls/ArtistEditionBurner.sol:57]: Should the message mention owner? We are presuming artist should do it?

File: contracts/v2/artist-controls/ArtistEditionBurner.sol [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    // if no tokens issued, simply disable the edition, burn it!
+    if (totalSupply == 0) {
+      kodaAddress.updateActive(_editionNumber, false);
+      emit EditionDeactivated(_editionNumber);
+    }
+    // if some tokens issued, reduce ths supply so that no more can be issued
+    else {
+      kodaAddress.updateTotalAvailable(_editionNumber, totalSupply);
+      emit EditionSupplyReduced(_editionNumber);
+    }
```
Comment (andygray) [inline contracts/v2/artist-controls/ArtistEditionBurner.sol:79]: 👍

File: contracts/v2/artist-controls/ArtistEditionBurner.sol [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    uint256 totalRemaining = kodaAddress.totalRemaining(_editionNumber);
+    require(totalRemaining > 0, "Only when edition not sold out");
+
+    // total issued so far
+    uint256 totalSupply = kodaAddress.totalSupplyEdition(_editionNumber);
+
+    // if no tokens issued, simply disable the edition, burn it!
+    if (totalSupply == 0) {
+      kodaAddress.updateActive(_editionNumber, false);
+      emit EditionDeactivated(_editionNumber);
```
Comment (andygray) [inline contracts/v2/artist-controls/ArtistEditionBurner.sol:73]: worth updating `totalAvailable` to zero here too?

File: test/truffle/v2/ArtistEditionBurner.test.js [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    beforeEach(async () => {
+      await this.koda.createActiveEdition(editionNumber2, editionData2, editionType, 0, 0, artistAccount, artistShare, edition2Price, editionTokenUri2, 3, {from: _owner});
+    });
+
+    it('fails if not called by artist', async () => {
+      await assertRevert(
+        this.burner.deactivateOrReduceEditionSupply(editionNumber2, {from: account2}),
+        "Only from the edition artist account"
+      );
+    });
```
Comment (andygray) [inline test/truffle/v2/ArtistEditionBurner.test.js:171]: add a test for contract owner to ensure we can access if needed?

File: contracts/v2/artist-controls/ArtistEditionBurner.sol (hunk 1 above)
Comment (jamesmorgan) [reply contracts/v2/artist-controls/ArtistEditionBurner.sol:57]: Ye presumed artist control - owner override as last resort

File: contracts/v2/artist-controls/ArtistEditionBurner.sol (hunk 3 above)
Comment (jamesmorgan) [reply contracts/v2/artist-controls/ArtistEditionBurner.sol:73]: Good shout - changing

**Pull request merged**